### DISCUSSION TITLE
Why data type indicator in uplink/downlink bytestream is in U32 format?

### DISCUSSION THREAD

**Discussion by rawinza555** [nasa/fprime#874] [Q&A] [answered]
I'm working on using NRF24L01 component on F' with the end goal to eventually replace TCP/IP and UDP for downlink and uplink data. One small issue I found is that RF24 only allows a payload with the size of 32 bytes. But telemetry data packet has 35 bytes so I take a look to see if I can make them fit in one transaction. Here's where Im quite confused. For the byte that indicate the type (1 for telemetry), U32 is used. Since there are only 3 types of packet so far for this framework, Im wondering if there's any necessity of using U32 here instead of U8 or U16 that I'm not aware of.

**Comment by LeStarch** [comment, marked as answer]
I believe you can set these in our configuration header.  They default to U32, but could be set smaller to reduce the size in serialized byte streams.  Note: at this time you'd need to modify fprime-gds to accept the new size of types (if you are using fprime-gds).

https://github.com/nasa/fprime/blob/48dec2b72386e8063f08e06285fde31eb0e0d87d/config/FpConfig.hpp#L57-L85

**Reply by timcanham** [reply]
The catch is that the size of the descriptor also has to be modified in the ground system where the packet is decoded, and it isn't parameterized in the same way.

**Reply by LeStarch** [reply]
Actually, the com type you refer to (Tlm/Cmd/Event) is set here: https://github.com/nasa/fprime/blob/48dec2b72386e8063f08e06285fde31eb0e0d87d/Svc/FramingProtocol/FprimeProtocol.hpp#L18

Note: this will change the size token and the start-word token.  Again, fprime-gds needs to mirror these changes if needed.

**Reply by LeStarch** [reply]
It is also possible to re-implement a framing protocol to frame any way you see fit.

**Reply by rawinza555** [reply]
Thanks! I thought there was something more behind the scene. I will definitely take a look but I'm most likely going to make some modification to the framing/deframing protocol to make it easier to splice or combine data from different transaction.
